Tracks: Don't track users in dev mode or when opted out

scruffian opened the pull request:
We shouldn't fire Tracks events when we are in dev mode, or if users haven't accepted our TOS, or if they have opted out of Tracks.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This PR adds a check for these three cases to ensure that we don't fire Tracks events in these cases

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This came up here: https://github.com/Automattic/jetpack/pull/13688#issuecomment-540026634, but is a more generic solution

#### Testing instructions:
* Upgrade your Jetpack site to Professional
* Add a Jetpack Search widget to the site
* Check that you see a `jetpack_search_widget_widget_added` event in Tracks
* Enable development mode (by adding this to `wp-config.php`):
`define( 'JETPACK_DEV_DEBUG', true );`
* Add a Jetpack Search widget to the site
* Check that you don't see another `jetpack_search_widget_widget_added` event in Tracks

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* No changelog required?

scruffian committed:
[not verified] Don't track users in dev mode or when opted out

jeherve committed:
Fix all phpcs issues

This will avoid having to bypass the pre-commit hook for all commits to this file

jeherve commented on packages/tracking/src/Tracking.php:
```diff
@@ -66,20 +89,26 @@ function record_user_event( $event_type, $data = array(), $user = null ) {
 
+		// Don't track users who have opted our, or not agreed to the TOS, or sites in development mode.
+		$status = new Status();
+		if ( ! \Jetpack::jetpack_tos_agreed() || ! empty( $_COOKIE['tk_opt-out'] ) || $status->is_development_mode() ) {
```
I don't think we need to be that detailed:
- Checking for the cookie isn't necessary as the opt-out is covered on WordPress.com. See p9SXHV-6Y-p2 to find out more.
- `jetpack_tos_agreed` checks if Jetpack is active, so I think that would cover development mode in most cases, except for folks that had Jetpack active, and thus gave their consent for tracking, and then activated development mode. I think that's fine though, as they gave their consent.

```suggestion
		if ( ! \Jetpack::jetpack_tos_agreed() ) {
```

kraftbj replied:
Removing the cookie check make sense. I'll push that up while I'm in there.

jeherve commented on packages/tracking/src/Tracking.php:
```diff
@@ -66,20 +89,26 @@ function record_user_event( $event_type, $data = array(), $user = null ) {
 		}
 
+		// Don't track users who have opted our, or not agreed to the TOS, or sites in development mode.
+		$status = new Status();
```
```suggestion
```

jeherve commented on packages/tracking/src/Tracking.php:
```diff
@@ -66,20 +89,26 @@ function record_user_event( $event_type, $data = array(), $user = null ) {
 			return false;
 		}
 
+		// Don't track users who have opted our, or not agreed to the TOS, or sites in development mode.
```
```suggestion
		// Don't track users who have not agreed to our TOS.
```

jeherve reviewed (changes requested):
I added a commit to fix all PHPCS issues, to avoid having to bypass the pre-commit hook for that file in the future.

I also have one remark, I think we could simplify this a bit.

kraftbj committed:
[not verified] Opt out of tracking in dev mode in legacy branch.

kraftbj committed:
PHPCS fixes

kraftbj committed:
Include development mode check within Jetpack::jetpack_tos_agreed.

The TOS check is used for confirming that a site is "Trackable". For sites in development mode or are not active, we should not track them.

kraftbj committed:
Docblock for is_active_and_not_development function

jeherve commented on packages/tracking/legacy/class.tracks-client.php:
```diff
@@ -38,22 +43,21 @@
 	 * @return mixed         True on success, WP_Error on failure
 	 */
-	static function record_event( $event ) {
-		if ( ! Jetpack::jetpack_tos_agreed() || ! empty( $_COOKIE['tk_opt-out'] ) ) {
```
I am going to walk back on some what I said in https://github.com/Automattic/jetpack/pull/13698#discussion_r333576946, sorry. It seems it was added on purpose in #9156, so maybe it's best if we keep it there?

Other than that it should be good to go.

jeherve replied:
done in 8a14c13

jeherve commented on packages/tracking/legacy/class.tracks-client.php:
```diff
@@ -38,22 +43,21 @@
-	static function record_event( $event ) {
-		if ( ! Jetpack::jetpack_tos_agreed() || ! empty( $_COOKIE['tk_opt-out'] ) ) {
+	public static function record_event( $event ) {
+		if ( ! Jetpack::jetpack_tos_agreed() ) {
```
```suggestion
		if ( ! Jetpack::jetpack_tos_agreed() || ! empty( $_COOKIE['tk_opt-out'] ) ) {
```

jeherve committed:
Bring cookie check back

Just to be sure. See https://github.com/Automattic/jetpack/pull/13698#discussion_r333960203

enejb commented on packages/tracking/src/Tracking.php:
```diff
@@ -66,20 +87,25 @@ function record_user_event( $event_type, $data = array(), $user = null ) {
 		}
 
+		// Don't track users who have opted out or not agreed to our TOS, or are not running an active Jetpack.
+		if ( ! \Jetpack::jetpack_tos_agreed() ) {
```
Since `Jetpack` is not a package this means that his package is not usable outside of jetpack. 
Which kind of defeats the purpose of this a package.

jeherve replied:
That's a good point. Isn't it already an issue with this already?
https://github.com/Automattic/jetpack/blob/8a14c13d207c4b3a9650df6f453f99579ed07835/packages/tracking/legacy/class.tracks-client.php#L60

Does this mean we should also refrain from relying on constants like `JETPACK__PLUGIN_FILE` or `JETPACK__VERSION`, that are not set anywhere but in Jetpack right now?

enejb replied:
To be honest I don't quite know. 
but I always assume that the `src` folder is there code that is supposed to be working also outside of jetpack and the `src` are classes that could be used outside of Jetpack. 

I think using a filter here instead would work well.

jeherve replied:
@enejb Is this what you had in mind? #13746

tyxla replied:
The constants are a different thing FWIW, we have the `constants` package and we can use it to access any constants inside packages without relying on the core plugin.
